ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcService.java

```diff
@@ -0,0 +1,652 @@
+
+  private HttpServer httpServer;
+  private final HealthService livenessService;
+  private final HealthService readinessService;
+
+  /**
+   * Construct a JsonRpcService to handle either http or websocket clients
+   *
+   * @param vertx The vertx process that will be running this service
+   * @param dataDir The data directory where requests can be buffered
+   * @param config Configuration for the rpc methods being loaded
+   * @param metricsSystem The metrics service that activities should be reported to
+   * @param natService The NAT environment manager.
+   * @param methods The json rpc methods that should be enabled
+   * @param maybeSockets websocket configuration to use
+   * @param scheduler for managing ETH tasks
+   * @param authenticationService used to grant or deny access to methods served
+   * @param livenessService A service responsible for reporting whether this node is live
+   * @param readinessService A service responsible for reporting whether this node has fully started
+   */
+  public JsonRpcService(
+      final Vertx vertx,
+      final Path dataDir,
+      final JsonRpcConfiguration config,
+      final MetricsSystem metricsSystem,
+      final NatService natService,
+      final Map<String, JsonRpcMethod> methods,
+      final Optional<WebSocketConfiguration> maybeSockets,
+      final EthScheduler scheduler,
+      final Optional<AuthenticationService> authenticationService,
+      final HealthService livenessService,
+      final HealthService readinessService) {
+    this.dataDir = dataDir;
+    this.requestTimer =
+        metricsSystem.createLabelledTimer(
+            BesuMetricCategory.RPC,
+            "request_time",
+            "Time taken to process a JSON-RPC request",
+            "methodName");
+    JsonRpcProcessor jsonRpcProcessor = new BaseJsonRpcProcessor();
+
+    this.socketConfiguration =
+        maybeSockets.isPresent() ? maybeSockets.get() : WebSocketConfiguration.createDefault();
+
+    if (authenticationService.isPresent()) {
+      jsonRpcProcessor =
+          new AuthenticatedJsonRpcProcessor(
+              jsonRpcProcessor,
+              authenticationService.get(),
+              this.socketConfiguration.getRpcApisNoAuth());
+    }
+
+    final JsonRpcExecutor jsonRpcExecutor = new JsonRpcExecutor(jsonRpcProcessor, methods);
+    this.webSocketMessageHandler =
+        Optional.of(
+            new WebSocketMessageHandler(
+                vertx, jsonRpcExecutor, scheduler, this.socketConfiguration.getTimeoutSec()));
+
+    validateConfig(config);
+    this.config = config;
+    this.vertx = vertx;
+    this.natService = natService;
+    this.rpcMethods = methods;
+    this.authenticationService = authenticationService;
+    this.livenessService = livenessService;
+    this.readinessService = readinessService;
+    this.maxActiveConnections = config.getMaxActiveConnections();
+  }
+
+  public CompletableFuture<?> start() {
```

Comment:
Consider `CompletableFuture<Void>` since the return type is never used.  Still uses `complete(null)` IIRC.